### ISSUE TITLE
file event callbacks

### ISSUE THREAD

**Issue by odygrd** [odygrd/quill#193]
### Discussed in https://github.com/odygrd/quill/discussions/183

<div type='discussions-op-text'>

<sup>Originally posted by **jmccabe** July 12, 2022</sup>
I've just been having a brief look at quill, with the focus on whether it could replace spdlog in our code because there are a couple of things we want to be able to do that, to some extent, are more difficult than we'd hoped with spdlog.

The first thing is that, when using a rotating file handler (or "sink" in spdlog terms), we'd like to log a few lines of "header" info in. Until recently, doing that in spdlog was awkward as there was no indication of when the rotate occurred, and the rotating file sink was a final class, with a private "rotate" function, that meant just deriving and handling the rotate slightly differently wasn't possible. Spdlog's latest version adds support for callbacks that help this to be done. Unfortunately, though, quill (unless I'm missing something) seems to suffer from the same restriction; class rotating_file_handler is private and there are no indications that a rotate is about to happen :-(

The other thing would be to take the formatted text and encode/encrypt it before it gets written to a file. While it looks like it may be possible to work around this, somehow (possibly in the same way you need to with spdlog currently), it would be really nice to be able to configure everything 'normally' then, when you're sure it's all working, just be able to pass a callback in where the formatted log message was passed in, and an encoded version of it was returned which was then stored in the file, instead of the original formatted string.

Are either of these the sort of thing that could be considered for quill?
</div>

**Comment by odygrd** [comment]
could have before_open, after_open , before_close , after_close for each opened file in the Handlers and possibly an `std::string on_write(log_message)` callback

**Comment by odygrd** [comment]
sorry for the late turnaround. This will be available in `v2.7.0`

Following callbacks are available

```c++
int main()
{
  quill::start();

  quill::FileEventNotifier fen;

  fen.before_open = [](quill::fs::path const& filename)
  { std::cout << "before opening " << filename << std::endl; };

  fen.after_open = [](quill::fs::path const& filename, FILE* f)
  { std::cout << "after opening " << filename << std::endl; };

  fen.before_close = [](quill::fs::path const& filename, FILE* f)
  { std::cout << "before closing " << filename << std::endl; };

  fen.after_close = [](quill::fs::path const& filename)
  { std::cout << "after closing " << filename << std::endl; };

  fen.before_write = [](quill::fmt_buffer_t const& formatted_log_message)
  { 
     // can be used for encode/encrypt as it will write result to the file
     quill::fmt_buffer_t result; 
     return result; 
  };

  quill::Handler* file_handler =
    quill::file_handler("myfile.log", "w", quill::FilenameAppend::None, std::move(fen));

  quill::Logger* mylogger = quill::create_logger("mylogger", file_handler);

  LOG_INFO(mylogger, "Hello world");
}
```
